doc/gmcpubar.1.txt
gmcpubar(1)
===========

NAME
----
gmcpubar - Graphical CPU bar for Dzen2

SYNOPSIS
--------
[verse]
'gmcpubar' [common options] [color options] [-f|--cpu=INDEX]

DESCRIPTION
-----------
gmcpubar produces CPU bar for Dzen2.  Many options can be used to tweak how exactly it should look.

OPTIONS
-------
Options for gmcpubar.

-f INDEX::
--cpu=INDEX::
        Index of the processor to watch.
        +
        If this is not specified, the bar shows overall CPU usage.  If this is specified, then a single logical processor usage is shown.

Color options for gmcpubar.

-a COLOR::
--kern=COLOR::
        Color for the kernel section of the bar.

-b COLOR::
--user=COLOR::
        Color for the user section of the bar.

-c COLOR::
--nice=COLOR::
        Color for the nice section of the bar.

-c COLOR::
--idle=COLOR::
        Color for the idle section of the bar.

Common options for all gm*bar commands.

-F COLOR::
--fg=COLOR::
        The outline color.  Default is "red".  If this is "none", outline is not drawn.

-B COLOR::
--bg=COLOR::
        The background color.  Default is "none" which means that background is not drawn.

-m PIXELS::
--margin=PIXELS::
        Number of pixels to leave between the edge of the window and the outline (which may be invisible).
        +
        Default is zero.  I.e. the outline resides directly next to window edges.

-p PIXELS::
--padding=PIXELS::
        Number of pixels to leave between outline (which may be invisible) and the sections of the bar.
        +
        Note that the padding is counted from "outer edge" of the outline.  This is deliberate to make it possible to draw the sections directly next to window edge by leaving both margin and padding to zero.
        +
        Padding defaults to zero.

-w PIXELS::
--width=PIXELS::
        Width of the bar in pixels.  This includes margins and paddings.

-h PIXELS::
--height=PIXELS::
        Height of the bar in pixels.  This includes margins and paddings.

-i SECONDS::
--interval=SECONDS::
        How ofter to refresh the bar, in seconds.

-L FILE::
--logfile=FILE::
        The name of the file for diagnostic messages.
        +
        Note that you can use same log file for multiple instances of gm*bar commands.

-P TEXT::
--prefix=TEXT::
        A snippet of text to output in front of the bar.
        +
        Note that you will have to take the width of the text into account when specifying window width for dzen2, but do not include the text width into --width above.

-S TEXT::
--suffix=TEXT::
        A snippet of text to output after the bar.
        +
        Note that you will have to take the width of the text into account when specifying window width for dzen2, but do not include the text width into --width above.

-s PIXELS::
--segment=PIXELS::
        This option, together with --gap, enables segmented drawing.  Both --segment and --gap must be greater than zero for this to make sense.
        +
        The default is to draw the sections in a continuous bar.
        +
        In segmented drawing, the sections are chopped into segments.  Note the the sections can end between segments or mid-segments, i.e. a color change can happen anywhere regardless of segments.
        +
        This option specifies the pixel width of a single segment.

-g PIXELS::
--gap=PIXELS::
        This option specifies how many pixels to leave between segments.  (See --segment.)

-G PIXELS::
--granularity=PIXELS::
        This specifies how many pixels at a time the sections grow or shrink.
        +
        The default zero has a special meaning: when calculating the section width, the width is always rounded down to even pixel width.  For example, if a section would occupy 50.9% of a 100 pixels, it would be rounded to 50 pixels.
        +
        By specifying granularity, the behaviour changes so that a more natural rounding is used.  In the preceding example, with --granularity=1, the width would become 51 pixels.  See --rounding below for more examples.
        +
        This option can be useful with --segment and --gap if you do not want the section to ever change in middle of a segment.  In this case, you would specify segment width N, gap width M, and granularity N+M.

-R FLOAT::
--rounding=FLOAT::
        When using --granularity, this specifies how the rounding should behave.  Without --granularity this has no effect.
        +
        For example, if you specify granularity to be 5 pixels and rounding to be 3, then a section with actual width 2 would be rounded to 5, while a section with actual width 1.999 would be rounded to zero.
        +
        The default rounding is half the granularity, which gives you the natural rounding (2.4 becomes 0, 2.5 becomes 5, in above example).

-V::
--version::
        Print version and exit with zero status.

--usage::
        Print usage and exit with zero status.

-?::
--help::
        Print help and exit with zero status.

FILES
-----

/proc/cpuinfo::
        This file is read once at the beginning of execution, to find out how many CPUs there are in the system.

/proc/stat::
        The source of CPU usage information.

AUTHOR
------
Original author and current maintainer: mailto:vmj@linuxbox.fi[Mikko Värri].

COPYRIGHT
---------
GNU GPLv3


src/gmcpubar.c
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

#include "libgmbar.h"
#include "common.h"
#include "log.h"
#include "readfile.h"
#include "buffer.h"
#include "version.h"

/* Maximum length of a CPU field label in /proc/stat ('cpu[CPU_INDEX] ') */
#define MAX_CPU_FIELD_LEN 20

/* Static functions */
static error_t handle_option(int key,
                             char* arg,
                             struct argp_state *state);
static int get_stat(buffer* stat,
                    const char* field,
                    unsigned int *kern,
                    unsigned int *user,
                    unsigned int *nice,
                    unsigned int *idle);
static long get_num_cpus();
static long parse_cpuinfo(const char* cpuinfo,
                          const unsigned int size);
static int parse_stat(const char* stat,
                      const unsigned int size,
                      const char* field,
                      unsigned int *kern,
                      unsigned int *user,
                      unsigned int *nice,
                      unsigned int *idle);
static unsigned int parse_unsigned_int(const char* str,
                                       const char** end);


/* Argp option keys (available: 'a'-'f') */
enum {
        OPTION_KERN_COLOR = 'a',
        OPTION_USER_COLOR = 'b',
        OPTION_NICE_COLOR = 'c',
        OPTION_IDLE_COLOR = 'd',
        OPTION_CPU_INDEX  = 'f',
};


/* Argp input */
typedef struct arguments arguments;
struct arguments {
        common_arguments common_config;
        int cpu_index;
};

/* Options */
static struct argp_option options[] = {
        { "kern",       OPTION_KERN_COLOR,         "COLOR",     0,
          "Color for the kernel portion of the bar"             },
        { "user",       OPTION_USER_COLOR,         "COLOR",     0,
          "Color for the user portion of the bar"               },
        { "nice",       OPTION_NICE_COLOR,         "COLOR",     0,
          "Color for the nice portion of the bar"               },
        { "idle",       OPTION_IDLE_COLOR,         "COLOR",     0,
          "Color for the idle portion of the bar"               },
        { "cpu",        OPTION_CPU_INDEX,          "INDEX",     0,
          "Index of the processor to watch"                     },
        { 0 }
};

/* Argp parser */
static const struct argp argp = { options, handle_option, NULL,
                                  "gmcpubar -- dzen2 cpu bar",
                                  common_argp_child
};

int
main(int argc, char** argv)
{
        int err = 0;
        long total; // Number of clock ticks per second
        long num_cpus;
        char cpu_field[MAX_CPU_FIELD_LEN + 1];
        unsigned int _kern, _user, _nice, _idle;
        unsigned int kern, user, nice, idle;
        arguments config;
        buffer* stat = NULL;
        gmbar* bar = NULL;

        stat = buffer_new();
        if (!stat)
        {
                return -1;
        }

        bar = gmbar_new_with_defaults(100, 10, "red", "none");
        if (!bar)
        {
                buffer_free(stat);
                return -1;
        }

        err = gmbar_add_sections(bar, 4, "red", "orange", "yellow", "none");
        if (err)
        {
                buffer_free(stat);
                gmbar_free(bar);
                return -1;
        }

        config.cpu_index = -1;
        config.common_config.bar = bar;
        config.common_config.interval = 15;
        config.common_config.prefix = NULL;
        config.common_config.suffix = NULL;
        err = argp_parse(&argp, argc, argv, 0, NULL, &config);
        if (err)
        {
                buffer_free(stat);
                gmbar_free(bar);
                return err;
        }

        /* Clock ticks per second (per CPU) */
        total = sysconf(_SC_CLK_TCK);

        /* Number of CPUs */
        num_cpus = get_num_cpus();
        if (num_cpus < 0)
        {
                buffer_free(stat);
                gmbar_free(bar);
                return num_cpus;
        }
        if (config.cpu_index >= 0
            && config.cpu_index <= CHAR_MAX
            && config.cpu_index < num_cpus)
        {
                snprintf(cpu_field, MAX_CPU_FIELD_LEN, "cpu%i ", config.cpu_index);
        }
        else
        {
                snprintf(cpu_field, MAX_CPU_FIELD_LEN, "cpu ");
        }

        /* Initialize history */
        err = get_stat(stat, cpu_field, &_kern, &_user, &_nice, &_idle);
        if (err)
        {
                buffer_free(stat);
                gmbar_free(bar);
                return err;
        }

        while (config.common_config.interval)
        {
                sleep(config.common_config.interval);

                err = get_stat(stat, cpu_field, &kern, &user, &nice, &idle);
                if (err)
                {
                        buffer_free(stat);
                        gmbar_free(bar);
                        return err;
                }

                /* total is not accurate */
                total = (kern - _kern) + (user - _user) + (nice - _nice) + (idle - _idle);

                gmbar_set_section_width(bar->sections[0], total, kern - _kern);
                gmbar_set_section_width(bar->sections[1], total, user - _user);
                gmbar_set_section_width(bar->sections[2], total, nice - _nice);
                gmbar_set_section_width(bar->sections[3], total, idle - _idle);

                err = print_bar(&config.common_config);
                if (err)
                {
                        buffer_free(stat);
                        gmbar_free(bar);
                        return err;
                }

                _kern = kern;
                _user = user;
                _nice = nice;
                _idle = idle;
        }

        return 0;
}

static error_t
handle_option(int key, char* arg, struct argp_state *state)
{
        error_t err = 0;
        arguments* config = (arguments*) state->input;

        switch (key)
        {
        case ARGP_KEY_INIT:
                state->child_inputs[0] = config;
                break;

        case OPTION_KERN_COLOR:
                err = parse_option_arg_string(arg, &config->common_config.bar->sections[0]->color);
                break;
        case OPTION_USER_COLOR:
                err = parse_option_arg_string(arg, &config->common_config.bar->sections[1]->color);
                break;
        case OPTION_NICE_COLOR:
                err = parse_option_arg_string(arg, &config->common_config.bar->sections[2]->color);
                break;
        case OPTION_IDLE_COLOR:
                err = parse_option_arg_string(arg, &config->common_config.bar->sections[3]->color);
                break;
        case OPTION_CPU_INDEX:
                config->cpu_index = parse_unsigned_int(arg, NULL);
                break;

        default:
                err = ARGP_ERR_UNKNOWN;
                break;
        }

        return err;
}

static int
get_stat(buffer* stat,
         const char* field,
         unsigned int *kern,
         unsigned int *user,
         unsigned int *nice,
         unsigned int *idle)
{
        int err = 0;

        *kern = *user = *nice = *idle = 0;

        /* reuse all the space */
        stat->len = 0;

        err = readfile("/proc/stat", &stat->buf, &stat->len, &stat->max);
        if (err)
        {
                return err;
        }

        err = parse_stat(stat->buf, stat->len, field, kern, user, nice, idle);
        return err;
}


static long
get_num_cpus()
{
        int err = 0;
        char* cpuinfo = NULL;
        unsigned int size = 0;
        unsigned int max = 0;

        err = readfile("/proc/cpuinfo", &cpuinfo, &size, &max);
        if (err || !cpuinfo)
        {
                return err;
        }

        return parse_cpuinfo(cpuinfo, size);
}

/**
 *
 */
static long
parse_cpuinfo(const char* cpuinfo, const unsigned int size)
{
        const char* p = cpuinfo;
        unsigned int len = size;
        long cpus = 0;

        do
        {
                p = memstr(p, "processor", len);
                if (p && (p == cpuinfo || p[-1] == '\n'))
                {
                        cpus++;
                        len = size - (p - cpuinfo);
                        p++;
                }
        } while (p);

        return cpus;
}

/**
 * Parse CPU field from stat.
 *
 * @param   stat      Contents of the /proc/stat file
 * @param   size      Size of the content
 * @param   field     Name of the field to parse, e.g. "cpu" or "cpu1", zero terminated
 * @param   kern      On return, contains the parsed value or zero.
 * @param   user      On return, contains the parsed value or zero.
 * @param   nice      On return, contains the parsed value or zero.
 * @param   idle      On return, contains the parsed value or zero.
 * @return  Zero on success, -1 if the field is not found.  Note that any
 * parse errors are not detected.
 */
static int
parse_stat(const char* stat,
           const unsigned int size,
           const char* field,
           unsigned int *kern,
           unsigned int *user,
           unsigned int *nice,
           unsigned int *idle)
{
        const char* p = stat;
        unsigned int len = size;

        /* Initialize to zeros */
        *kern = *user = *nice = *idle = 0;

        /* Find the field */
        do
        {
                p = memstr(p, field, len);
                if (!p)
                {
                        log_error("Field not found: %d", -1);
                        return -1;
                }
        } while (p != stat && p[-1] != '\n' && p++ && (len = size - (p - stat)));

        /* Skip the label */
        p += strlen(field);

        *user = parse_unsigned_int(p, &p);
        *nice = parse_unsigned_int(p, &p);
        *kern = parse_unsigned_int(p, &p);
        *idle = parse_unsigned_int(p, NULL);

        return 0;
}

/**
 * @param   str   String to parse
 * @return  Parsed value.
 */
static unsigned int
parse_unsigned_int(const char* str, const char** end)
{
        unsigned int value = 0;
        while (!isdigit(*str))
                str++;
        while(isdigit(*str))
                value = value * 10 + (*str++ - '0');
        if (end)
                *end = (char*)str;
        return value;
}
